### PR TITLE
feat: [WIP/Draft] add optional flash attn in attention block + test

### PR DESCRIPTION
**Pull request by art-test-stack**
Hi,

This pull request adds an optional flash attention support with kernel "varunneal/flash-attention-3"
Theoretically, it should run faster if the hardware supports flash-attn. 
However, this is quite WIP, i am not sure "flash_attn" option should be in AttentionConfig, but rather on a higher level (it should not change model behavior, it is a "running" config).
I don't have cuda device currently to properly test it, I hope this helps though

### CODE REVIEW

**Comment by art-test-stack** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by art-test-stack** [comment]
recheck

**Comment by Ndles** [comment]
hi @art-test-stack , thanks for creating this PR! We will take a look at it some time next week and give feedback then ✌️

**Comment by art-test-stack** [comment]
hi @Ndles, thanks for your attention. 
For now, I mainly need to know first if you were considering this kind of improvement (have not found a similar idea/decision/thought on it in the codebase); or if there were some specific reason why it was not implemented yet, or if you had another idea instead.
However, it is still quite WIP. I just found other places in the codebase that could benefit from the same kind. 
I will open for review today or over the weekend, when I will have pushed changes, where it seems to be nice to use flash_attn kernels 😊

EDIT: actually I am facing a wall regarding the mask that patches the resolution in attn mask (in vit, for example) as flash attn does not handle custom masks 😅 
I am not familiar with vision transformer, but I am curious to see if the logic can be reimplemented so it can work

**Comment by Ndles** [comment]
hey @art-test-stack 

Thanks for picking this up! Straight answer to your main question: no decision against faster kernels, it just hadn't been done - so yes, we want this.
  
You're right on the big one - `use_flash_attn` doesn't belong on `AttentionConfig`. That config describes what the model is and gets baked into checkpoints; kernel choice is just how you run it on a given GPU. Let's make it a run-level setting (something like `attention_backend` on the root config), resolved once at run setup rather than at import - the current import-time Hub download will bite people in CI/offline. And ideally one shared dispatch helper that all the SDPA modules route through, not a flag on one module, since Perceiver/Transolver/anchor-attention all use SDPA too.

With that implemented it should be possible to use FA3 and FA4 (hopefully with minimal tweaks 🤞 ).

We will review your suggestion when you're ready :) 

One note: don't forget to update pyproject.toml and uv.lock when needed.

File: src/noether/modeling/modules/attention/dot_product.py [hunk 1]
```diff
@@ -91,9 +132,18 @@ def forward(
...
             assert freqs is None
 
-        x = F.scaled_dot_product_attention(
-            q, k, v, attn_mask=attn_mask, dropout_p=self.dropout if self.training else 0.0
-        )
+        if self.use_flash_attn and attn_mask is None and _FLASH_ATTN3 is not None:
+            x = _FLASH_ATTN3.flash_attn_func(
+                q.transpose(1, 2),
+                k.transpose(1, 2),
+                v.transpose(1, 2),
```
Comment (HennerM) [inline src/noether/modeling/modules/attention/dot_product.py]: We should be able to remove the transpose operations if we change the rearrange above and below.

File: src/noether/modeling/modules/attention/dot_product.py [hunk 2]
```diff
@@ -10,10 +13,47 @@
...
+    flash_attn_interface = getattr(kernel, "flash_attn_interface", None)
+    if flash_attn_interface is None:
+        return None
+    if not hasattr(flash_attn_interface, "flash_attn_func"):
+        return None
+
+    return flash_attn_interface
+
+
+_FLASH_ATTN3 = _load_flash_attn3_interface()
```
Comment (HennerM) [inline src/noether/modeling/modules/attention/dot_product.py]: this will download and load FA3 kernel regardless of it being actually used given the config.

File: src/noether/core/schemas/modules/attention.py [hunk 3]
```diff
@@ -54,6 +54,9 @@ class AttentionConfig(BaseModel):
     qk_norm: bool = Field(False)
     """Whether to apply layer normalization to the query and key features before computing attention scores."""
 
+    use_flash_attn: bool = False
```
Comment (HennerM) [inline src/noether/core/schemas/modules/attention.py]: This is slightly misleading since even if this is False, we are using flash attention through torch's scaled_dot_product_attention. I think it would be more flexible to configure the implementation with a string, similar to what axolotl is doing: https://docs.axolotl.ai/docs/attention.html that way we can change between FA4/FA3 and perhaps even arbitrary attention implementations.

**Comment by HennerM** [comment]
IMO I think a `implementation` config makes most sense in the `AttentionConfig`. Whie we could argue that this is more of an environment setup, all of our config is done this way so I think keeping it there is the most consistent.

> EDIT: actually I am facing a wall regarding the mask that patches the resolution in attn mask (in vit, for example) as flash attn does not handle custom masks 😅

We don't need to support all kernels for all models. I think it's fine to raise a validation error in case someone selects an implementation that doesn't support a custom mask (although I am a bit surprised that not all kernels support this tbh).

**Review by HennerM — COMMENTED** [review]
Thanks for the contribution. I think the cleanest way to add this in is get inspired by how Axolotl is doing this: https://docs.axolotl.ai/docs/attention.html
